internal/manager/run.go

```diff
@@ -234,6 +206,106 @@ func Run(ctx context.Context, c *Config, diagnostic util.ConfigDumpDiagnostic, d
 	return mgr.Start(ctx)
 }
 
+func validateKongRoots(kongRoots []map[string]any, skipCACerts bool) (string, kong.Version, error) {
+	var (
+		errs        []error
+		kongVersion kong.Version
+		dbMode      string
+	)
+
+	uniqs := lo.UniqBy(kongRoots, func(kongRoot map[string]any) string {
+		kongRootConfig, ok := kongRoot["configuration"].(map[string]any)
+		if !ok {
+			errs = append(errs,
+				fmt.Errorf("invalid root configuration, expected a map[string]any got %T",
+					kongRoot["configuration"]),
+			)
+			return ""
+		}
+
+		dbMode, ok = kongRootConfig["database"].(string)
+		if !ok {
+			errs = append(errs,
+				fmt.Errorf("invalid database configuration, expected a string got %t",
+					kongRootConfig["database"]),
+			)
+			return ""
+
+		}
+
+		if dbMode == "off" && skipCACerts {
+			errs = append(errs,
+				fmt.Errorf("--skip-ca-certificates is not available for use with DB-less Kong instances"),
+			)
+			return ""
+		}
+
+		var err error
+		kongVersion, err = kong.ParseSemanticVersion(kong.VersionFromInfo(kongRoot))
+		if err != nil {
+			errs = append(errs,
+				fmt.Errorf("could not parse Kong version, version-specific behavior disabled"),
+			)
+			return ""
+		}
+
+		version := kong.VersionFromInfo(kongRoot)
+		return fmt.Sprintf("%s:%s", dbMode, version)
```

Comment:
ditto - separate function? I think that we might also consider separating the validation part from the uniqueness check - it's quite confusing to me that we do both in a single iteration. I understand that can be more performant, but IMO it would be easier to read something like 
```go
if errs := lo.Map(kongRoots, validateKongRoot); len(errs) > 0 {
...
}
if uniqs := lo.UniqBy(kongRoots, kongRootKey); len(uniqs) > 1 {
...
}
```
Would that make sense?

Reply:
I probably went a little over the top but there's https://github.com/Kong/kubernetes-ingress-controller/blob/5317416fc9468a0c6e7b6a88dd371558e8ff4e16/internal/manager/utils/kongroot.go now to handle this and it does what you described - I believe. Let me know if it makes sense.

Reply:
Yup, that even exceeds my expectations :) Nice!